### PR TITLE
feat: in-mem implementation and testsuite

### LINKED ISSUE #3172: Define the core abstractions for the storage package
Define the core chunk related abstractions

Define the data abstraction layer interfaces for indexes.

### PR DESCRIPTION
**Pull request by aloknerurkar**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [x] My change requires a documentation update and I have done it
- [x] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues

### Description
This PR implements the in-mem Store. This is part of the localstore rewrite epic.

### Related Issue (Optional)
The PR closes #3172 #3173 #3174 #3180 

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3190)

### CODE REVIEW

File: pkg/storagev2/testsuite/testsuite.go [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+		t.Fatalf("failed marshaling %s", err.Error())
+	}
+
+	buf2, err := b.Marshal()
+	if err != nil {
+		t.Fatalf("failed marshaling %s", err.Error())
+	}
+
+	if bytes.Compare(buf1, buf2) != 0 {
+		t.Fatal("bytes not equal for item %s/%s", a.Namespace(), a.ID())
```
Comment (aloknerurkar) [inline pkg/storagev2/testsuite/testsuite.go:87]: Fatalf

**Review by mrekucci — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3190)* status: all files reviewed, 1 unresolved discussion (waiting on @mrekucci)

___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 1 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VMsfy7_QpEiW0RSx9:-N9VMsfy7_QpEiW0RSxA:b-zhdpvi) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L1)):*
> ```Go
> package inmem
> ```

Missing header.
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 20 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VNbua40ixMJAdH6IM:-N9VNbua40ixMJAdH6IN:b-hsg5tx) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L20)):*
> ```Go
> // store. So if we use the go in-built map, we might end up complicating this implementation.
> // Instead we will use the hashicorp/go-radix implementation. This pkg provides a mutable radix
> // which gives O(k) lookup and ordered iteration
> ```

This is an unnecessary implementation detail. Just mentioning using the radix tree instead of the map should be enough. E.g.: "Since we need a lexicographic arrangement of keys for the store, this implementation uses a radix tree instead of a map. This makes the implementation much simpler and maintainable with O(k) lookup and ordered iteration".



_Suggestion:_
```Go
// ... We need lexicographic ordering of keys for the
// store. So if we use the go in-built map, we might end up complicating this implementation.
// Instead we will use the hashicorp/go-radix implementation. This pkg provides a mutable radix
// which gives O(k) lookup and ordered iteration
```
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 50 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VPRHv3gNk7ECLHhEi:-N9VPRHv3gNk7ECLHhEj:bqpgvw7) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L50)):*
> ```Go
> 	err := i.Unmarshal(val.([]byte))
> 	if err != nil {
> 		return fmt.Errorf("failed unmarshaling item %w", err)
> ```

Adding `:` before the `%w` would make it more readable.
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 84 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VPxVOEVQdpsamTlKu:-N9VPxVOEVQdpsamTlKv:bf26u1) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L84)):*
> ```Go
> 	val, err := i.Marshal()
> 	if err != nil {
> 		return fmt.Errorf("failed marshaling item %w", err)
> ```

The same as above.
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 120 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VQQ0O9w8tHCpdz1h2:-N9VQQ0O9w8tHCpdz1h3:b-b2rgbe) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L120)):*
> ```Go
> func (s *store) Iterate(q storage.Query, fn storage.IterateFn) error {
> 	if err := q.Validate(); err != nil {
> 		return fmt.Errorf("invalid query %w", err)
> ```

The `Validate` already returns "invalid query" string. To avoid stuttering, I would suggest returning something like: `fmt.Errorf("iteration failed: %w", err)`.
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 142 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VRjjX3PPTL2n1Et45:-N9VRjjX3PPTL2n1Et46:bbnybev) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L142)):*
> ```Go
> 			err := newItem.Unmarshal(v.([]byte))
> 			if err != nil {
> 				return nil, fmt.Errorf("failed unmarshaling %w", err)
> ```

Missing `:`.
___
*[`pkg/storagev2/inmemstore/inmemstore.go` line 161 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VS4WECYr9BlX90c5z:-N9VS4WF6j5S11S81jxu:bf26u1) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore.go#L161)):*
> ```Go
> 				stop, err := fn(*res)
> 				if err != nil {
> 					retErr = multierror.Append(retErr, fmt.Errorf("failed in iteration %w", err))
> ```

The same as above.
___
*[`pkg/storagev2/inmemstore/inmemstore_test.go` line 1 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VSmxL0TUXDMf-vTZP:-N9VSmxL0TUXDMf-vTZQ:b-zhdpvi) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/inmemstore/inmemstore_test.go#L1)):*
> ```Go
> package inmem_test
> ```

Missing header.
___
*[`pkg/storagev2/testsuite/testsuite.go` line 1 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VSr2k1_3hTlXKrpz2:-N9VSr2k1_3hTlXKrpz3:b-zhdpvi) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/testsuite/testsuite.go#L1)):*
> ```Go
> package storetesting
> ```

Missing header.
___
*[`pkg/storagev2/testsuite/testsuite.go` line 151 at r2](https://reviewable.io/reviews/ethersphere/bee/3190#-N9VTrTEEUjeIUm70fBL:-N9VTrTEEUjeIUm70fBM:b-qqst0v) ([raw file](https://github.com/ethersphere/bee/blob/5f5cf7302068374da06209364de2350d42bccc16/pkg/storagev2/testsuite/testsuite.go#L151)):*
> ```Go
> 			err := s.Put(i)
> 			if err != nil {
> 				t.Fatalf("failed to add new entry: %s", err.Error())
> ```

No need to call `Error()` when you use `%v` instead of `%s`. The same is in the test cases below.

**Review by mrekucci — APPROVED** [review]
Apart from the comments LGTM.

**Review by notanatol — APPROVED** [review]

File: pkg/storagev2/inmemstore/inmemstore.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+// license that can be found in the LICENSE file.
+
+package inmem
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/armon/go-radix"
```
Comment (istae) [inline pkg/storagev2/inmemstore/inmemstore.go:12]: is this the hashicorp one?

**Review by istae — APPROVED** [review]
This is great!

The iterator function could be greatly simplified if we just returned the key and byte value pair for the callback. It would get rid of the factory, item attribute, etc.... I understand that it passes the responsibility to the user to do the unmarshalling but it simplifies everything else.

**Comment by aloknerurkar** [comment]
> The iterator function could be greatly simplified if we just returned the key and byte value pair for the callback. It would get rid of the factory, item attribute, etc.... I understand that it passes the responsibility to the user to do the unmarshalling but it simplifies everything else.

So this would add up I think. For each instance in the codebase if we do Unmarshal and the error check, it would add a lot of unnecessary code. This is unnecessary as the store already knows how to interpret the bytes because of the Serializable interface.

This could be improved later once we have generics. We can get rid of the factory if we just infer the type from the method itself.

You can get access to the bytes directly using the filters. This would allow you to prevent unnecessary unmarshaling if you know can check on the bytes directly. There is a test case for this as well.

File: pkg/storagev2/inmemstore/inmemstore.go (hunk 2 above)
Comment (aloknerurkar) [reply pkg/storagev2/inmemstore/inmemstore.go:12]: Yes. Hashicorp has a immutable-radix implementation. In that they have linked this as the mutable version. This is what we need. Its a very good and stable library I think.

**Review by janos — APPROVED** [review]
👍

**Pull request merged**